runtime/src/main/java/org/corfudb/util/NetworkUtils.java

```diff
@@ -27,20 +38,75 @@
-                if (currentAddress instanceof Inet4Address) {
-                    return currentAddress.getHostAddress();
+            for (NetworkInterface networkInterface : getAllInterfaces()) {
+                if (networkInterface.isUp()
+                        && interfaceName.equals(networkInterface.getName())) {
+                    // If the interface is a subinterface return any ipv4 address
+                    String address;
+                    if (networkInterface.isVirtual()) {
+                        address = getIPV4Address(networkInterface.getInterfaceAddresses());
+                    } else {
+                        // This is not a subinterface, need to exclude the subinterface addresses
+                        // from the address list before we select an ipv4 address.
+                        Set<InterfaceAddress> parentAddresses = new HashSet<>(networkInterface.getInterfaceAddresses());
+                        Set<InterfaceAddress> subInterfacesAddresses = getSubInterfacesAddresses(networkInterface);
+                        parentAddresses.removeAll(subInterfacesAddresses);
+                        address = getIPV4Address(new ArrayList<>(parentAddresses));
+                    }
+
+                    if (address != null) {
+                        return address;
+                    }
+
+                    break;
                 }
             }
         } catch (SocketException e) {
             log.error("Error fetching address from interface name ", e);
             throw new UnrecoverableCorfuError(e);
         }
+
         throw new UnrecoverableCorfuError("No valid interfaces with name "
                 + interfaceName + " found.");
     }
+
+    /**
+     * Collects all subinterface addreses for an interface
+     */
+    private static Set<InterfaceAddress> getSubInterfacesAddresses(NetworkInterface networkInterface) {
+        Set<InterfaceAddress> interfaceAddresses = new HashSet<>();
+        for (NetworkInterface subInterface : Collections.list(networkInterface.getSubInterfaces())) {
+            interfaceAddresses.addAll(subInterface.getInterfaceAddresses());
+        }
+        return interfaceAddresses;
+    }
+
+    /**
+     * Finds an ipv4 address in a list of interface addresses
+     */
+    private static String getIPV4Address(List<InterfaceAddress> addresses) {
+        for (InterfaceAddress interfaceAddress : addresses) {
+            if (interfaceAddress.getAddress() instanceof Inet4Address) {
+                return interfaceAddress.getAddress().getHostAddress();
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Get all interfaces, including virtual interfaces.
+     * @return A list of all interfaces
+     * @throws SocketException
+     */
+    private static List<NetworkInterface> getAllInterfaces() throws SocketException {
+        List<NetworkInterface> allInterfaces = Collections.list(NetworkInterface.getNetworkInterfaces())
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Immediately return this expression instead of assigning it to the temporary variable "allInterfaces". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1488)